hack/lib/golang.sh

```diff
@@ -521,7 +521,10 @@ kube::golang::setup_env() {
   kube::golang::create_gopath_tree
   export GOPATH="${KUBE_GOPATH}"
 
-  export GOCACHE="${KUBE_GOPATH}/cache"
+  # If these are not set, set them now.  THis ensures that any subsequent
+  # scripts we call (which may call this function again) use the same values.
+  export GOCACHE="${GOCACHE:-"${KUBE_GOPATH}/cache/build"}"
+  export GOMODCACHE="${GOMODCACHE:-"${KUBE_GOPATH}/cache/mod"}"
```

Comment:
Yeah, this is kind of a mess because of a lot of reasons.  Until workspaces lands, we are forced to phony-up a GOPATH.  Go's default is to put the module cache under GOPATH[0], which means it gets stored in this directory tree.

Running in a work tree and then running one of these scripts means the GOPATH changes and the modcache is found to be empty, so it re-downloads.

This change explicitly points the caches to the "real" tree, so we avoid the re-download, but as you point out, it is not very well managed.

I wish I could say workspaces fixes this, but it doesn't.  That tree (for now) STILL sets up a phony GOPATH, though for a different reason - `go install`.  We can't just set GOBIN (not compatible with crossbuilds).  We *could* use `go build -o` but then EVERYONE needs to get it just right (and a trailing slash matters, if I recall).  So the only way I found to make it easy to build tools was to set GOPATH and use `go install`, which writes to GOPATH[0]/bin.  PUUUUUUUKE.  I'd much rather run with GOPATH=/dev/null.  Better ideas are welcome :)